### ISSUE TITLE
Updated Ansible, and my dicts are no longer merging

### ISSUE THREAD

**Issue by jerrac** [ansible/ansible#10406]
Every time I run a playbook, I get this error:

```
ERROR: failed to combine variables, expected dicts but got a 'dict' and a 'NoneType'
```

I have `hash_behaviour = merge` set in my ansible.cfg. If I switch to `hash_behaviour = replace` my playbook runs.

As far as I can see, I have no syntax errors in my various var files.

This only popped up in the past couple days after I updated to the latest devel code.

```
ansible --version
ansible 1.9 (devel fb717ffd6c) last updated 2015/03/06 15:47:53 (GMT -700)
  lib/ansible/modules/core: (detached HEAD b0bc6f1379) last updated 2015/03/06 15:47:58 (GMT -700)
      lib/ansible/modules/extras: (detached HEAD 46e316a20a) last updated 2015/03/06 15:47:58 (GMT -700)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/03/06 15:47:58 (GMT -700)
  v2/ansible/modules/extras: (detached HEAD 46e316a20a) last updated 2015/03/06 15:47:58 (GMT -700)
  configured module search path = None
```

I'm on Ubuntu 14.04 64bit. Running ansible from the git repo.

Suggestions?

**Comment by jerrac** [comment]
I just updated to the latest devel code, and it still does this.

```
ansible --version
ansible 1.9 (devel 70f56c135c) last updated 2015/03/13 12:02:51 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 31cc5f543f) last updated 2015/03/13 12:03:08 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 8baba98ebe) last updated 2015/03/13 12:03:08 (GMT -700)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/03/13 12:03:08 (GMT -700)
  v2/ansible/modules/extras: (detached HEAD 46e316a20a) last updated 2015/03/13 12:03:08 (GMT -700)
  configured module search path = None
```

So I tried switching to v1.8.4 with a `git checkout v1.8.4 ; git submodule update --init` and this is what I got:

```
ansible --version               
ansible 1.8.4 (detached HEAD ebc8d48d34) last updated 2015/03/13 12:25:09 (GMT -700)
  lib/ansible/modules/core: (detached HEAD f22df78345) last updated 2015/03/13 12:25:16 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 23190986fd) last updated 2015/03/13 12:25:16 (GMT -700)
  v2/ansible/modules/core: (detached HEAD cb69744bce) last updated 2015/03/13 12:25:16 (GMT -700)
  v2/ansible/modules/extras: (detached HEAD 8a4f07eecd) last updated 2015/03/13 12:25:16 (GMT -700)
  configured module search path = None

ansible-playbook -i hosts --sudo --limit="hostname" apply.yml
ERROR: failed to combine variables, expected dicts but got a 'dict' and a 'NoneType'
```

Switched to v1.7.2 and received the following:

```
ansible --version
ansible 1.7.2
ansible-playbook -i hosts --sudo --limit="hostname" apply.yml
 [WARNING]: It is unnecessary to use '{{' in loops, leave variables in loop
expressions bare.

Traceback (most recent call last):
  File "/home/localusername/vendor/ansible/bin/ansible-playbook", line 309, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/localusername/vendor/ansible/bin/ansible-playbook", line 249, in main
    pb.run()
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/__init__.py", line 305, in run
    play = Play(self, play_ds, play_basedir, vault_password=self.vault_password)
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/play.py", line 157, in __init__
    self._tasks      = self._load_tasks(self._ds.get('tasks', []), load_vars)
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/play.py", line 558, in _load_tasks
    default_vars = utils.combine_vars(self.default_vars, default_vars)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 1281, in combine_vars
    return merge_hash(a, b)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 715, in merge_hash
    result[k] = merge_hash(a[k], v)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 710, in merge_hash
    for k, v in dicts.iteritems():
AttributeError: 'NoneType' object has no attribute 'iteritems'
```

After finding the with_dict calls that had '{{' in them, and fixing them, I get this:

```
ansible-playbook -i hosts --sudo --limit="hostname" apply.yml
Traceback (most recent call last):
  File "/home/localusername/vendor/ansible/bin/ansible-playbook", line 309, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/localusername/vendor/ansible/bin/ansible-playbook", line 249, in main
    pb.run()
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/__init__.py", line 305, in run
    play = Play(self, play_ds, play_basedir, vault_password=self.vault_password)
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/play.py", line 157, in __init__
    self._tasks      = self._load_tasks(self._ds.get('tasks', []), load_vars)
  File "/home/localusername/vendor/ansible/lib/ansible/playbook/play.py", line 558, in _load_tasks
    default_vars = utils.combine_vars(self.default_vars, default_vars)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 1281, in combine_vars
    return merge_hash(a, b)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 715, in merge_hash
    result[k] = merge_hash(a[k], v)
  File "/home/localusername/vendor/ansible/lib/ansible/utils/__init__.py", line 710, in merge_hash
    for k, v in dicts.iteritems():
AttributeError: 'NoneType' object has no attribute 'iteritems'
```

Could any of this have to do with what version of Python I'm on? The host I'm testing this on is running Ubuntu 14.04 LTS, and so is my local machine. Python is 2.7.6.

**Comment by jerrac** [comment]
Wow. The issue was a variable in the defaults file of a role I hadn't touched since 12/19/2014.

I had an example variable in the defaults file, with the actual value commented out. I had forgotten to initiate the dictionary with a `{}`. Apparently, before a week or two ago, merging somehow ignored that error. 

To clarify my situation a bit, I have a common role that depends on the role with the error. I've been running that role over and over and over again without the "NoneType" error for a long time. So when it suddenly showed up, and I hadn't made any changes recently, I was rather confused. Took me a bit to figure out I needed to test each role that common depended on.

Ah well, at least now it's fixed.

Would it be possible to change the error message:

```
ERROR: failed to combine variables, expected dicts but got a 'dict' and a 'NoneType'
```

to something like:

```
ERROR: failed to combine variables path/to/var/file var_name, and path/to/var/file/ var_name, expected dicts but got a 'dict' and a 'NoneType'
```

Basically, add information to the error message about what variable is coming through as 'NoneType'.

Should I post that idea as a separate issue?
